internal/handlers/sqlite/msg_aggregate.go

```diff
@@ -16,13 +16,335 @@ package sqlite
 
 import (
 	"context"
+	"errors"
+	"fmt"
+	"math"
+	"time"
 
+	"github.com/FerretDB/FerretDB/internal/backends"
+	"github.com/FerretDB/FerretDB/internal/clientconn/conninfo"
+	"github.com/FerretDB/FerretDB/internal/clientconn/cursor"
+	"github.com/FerretDB/FerretDB/internal/handlers/common"
+	"github.com/FerretDB/FerretDB/internal/handlers/common/aggregations"
+	"github.com/FerretDB/FerretDB/internal/handlers/common/aggregations/stages"
+	"github.com/FerretDB/FerretDB/internal/handlers/commonerrors"
+	"github.com/FerretDB/FerretDB/internal/handlers/commonparams"
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/iterator"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
 	"github.com/FerretDB/FerretDB/internal/util/must"
 	"github.com/FerretDB/FerretDB/internal/wire"
 )
 
 // MsgAggregate implements HandlerInterface.
 func (h *Handler) MsgAggregate(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg, error) {
-	// TODO https://github.com/FerretDB/FerretDB/issues/3148
-	return nil, notImplemented(must.NotFail(msg.Document()).Command())
+	document, err := msg.Document()
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	common.Ignored(document, h.L, "lsid")
+
+	if err = common.Unimplemented(document, "explain", "collation", "let"); err != nil {
+		return nil, err
+	}
+
+	common.Ignored(
+		document, h.L,
+		"allowDiskUse", "bypassDocumentValidation", "readConcern", "hint", "comment", "writeConcern",
+	)
+
+	var db string
+
+	if db, err = common.GetRequiredParam[string](document, "$db"); err != nil {
+		return nil, err
+	}
+
+	collectionParam, err := document.Get(document.Command())
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO handle collection-agnostic pipelines ({aggregate: 1})
+	// https://github.com/FerretDB/FerretDB/issues/1890
+	var ok bool
+	var collection string
+
+	if collection, ok = collectionParam.(string); !ok {
+		return nil, commonerrors.NewCommandErrorMsgWithArgument(
+			commonerrors.ErrFailedToParse,
+			"Invalid command format: the 'aggregate' field must specify a collection name or 1",
+			document.Command(),
+		)
+	}
+
+	dbPool, err := h.b.Database(db)
+	if err != nil {
+		if backends.ErrorCodeIs(err, backends.ErrorCodeDatabaseNameIsInvalid) {
+			msg := fmt.Sprintf("Invalid namespace specified '%s.%s'", db, collection)
+			return nil, commonerrors.NewCommandErrorMsgWithArgument(commonerrors.ErrInvalidNamespace, msg, document.Command())
+		}
```

Comment:
Can we add a test for that, or it is not possible?

Reply:
Not in this issue/PR; we will have time later to cleanup the handler once we have the only one

Reply:
In a couple of weeks time, all these codes will be moved elsewhere, we can test them that time :hugs: